caldav/davclient.py

```diff
@@ -539,14 +539,38 @@ def request(self, url, method="GET", body="", headers={}):
             auth_type = r.headers['WWW-Authenticate']
             auth_type = auth_type[0:auth_type.find(" ")]
 
+            error.assert_(auth_type in ('Basic', 'Digest'))
+
             if auth_type == 'Basic':
                 self.auth = requests.auth.HTTPBasicAuth(self.username, self.password)
             elif auth_type == 'Digest':
                 self.auth = requests.auth.HTTPDigestAuth(self.username, self.password)
             else:
                 raise NotImplementedError("Auth method %s not supported yet" % auth_type)
-            del self.username
```

Comment:
I thought it should be set to None?  I don't remember why I put in those lines right now, but there was a reason for it ... hm.

Reply:
Ah, nevermind - I see it's done further below

Reply:
Removing these two lines altogether (and not yet setting ```self.password``` and ```self.username``` to ```None```) is necessary, since both values are needed for retry in the next iteration of the call in the next ```elif ...``` branch.
By the way, it took me a while to realize that you deal with authorization check and final authentication by recursively calling ```self.request()```, with implicit state tracking in ```self.auth``` and ```self.username```.  Elegant but a bit unintuitive.

Reply:
If it wasn't intuitive, then probably it wasn't elegant :-)  (but at least better than the approach in v0.8.2).

I'm quite annoyed spending so much time on this - I think this logic should have been in the requests library, as it is neither DAV-specific nor CalDAV-specific.